src/main/java/org/elasticsearch/search/highlight/HighlightPhase.java

```diff
@@ -99,10 +102,10 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
 
                 String highlighterType = field.fieldOptions().highlighterType();
                 if (highlighterType == null) {
-                    boolean useFastVectorHighlighter = fieldMapper.fieldType().storeTermVectors() && fieldMapper.fieldType().storeTermVectorOffsets() && fieldMapper.fieldType().storeTermVectorPositions();
+                    boolean useFastVectorHighlighter = highlighters.get("fvh").canHighlight(fieldMapper);
                     if (useFastVectorHighlighter) {
                         highlighterType = "fvh";
-                    } else if (fieldMapper.fieldType().indexOptions() == IndexOptions.DOCS_AND_FREQS_AND_POSITIONS_AND_OFFSETS) {
+                    } else if (highlighters.get("postings").canHighlight(fieldMapper)) {
```

Comment:
It does seem a bit cleaner to use a list.

For what its worth my highlighter would just ```return true``` from canHighlight because its defaults let it pick its hit detection strategy based on what is in the mapping.